AGENTS.md

```diff
@@ -0,0 +1,109 @@
+
+### Build System
+
+The project uses:
+
+- **dune**: For building OCaml components (analysis & tools)
+- **esbuild**: For bundling TypeScript client/server
+- **npm**: For JavaScript/TypeScript dependencies
+- **TypeScript**: For type checking the client/server code
+
+## Development Commands
+
+### Initial Setup
+
+```bash
+npm install                    # Install all dependencies including client/server
+opam switch 5.2.0             # Install OCaml switch (if not already done)
+opam install . --deps-only    # Install OCaml dependencies
+```
+
+### Building
+
+```bash
+make build                    # Build OCaml binaries and copy to root
+npm run compile               # Compile TypeScript (client & server)
+npm run bundle                # Bundle for production (esbuild)
+npm run vscode:prepublish     # Clean and bundle (used for publishing)
+```
+
+### Development
+
+```bash
+npm run watch                 # Watch TypeScript compilation
+make -C analysis test         # Run analysis tests
+make -C tools/tests test      # Run tools tests
+make test                     # Run all tests
+```
+
+### Code Quality
+
+```bash
+make format                   # Format OCaml (dune) and JS/TS (prettier)
+make checkformat              # Check formatting without modifying
+make clean                    # Clean build artifacts
+```
+
+### Running the Extension in Development
+
+1. Open the project in VSCode
+2. Press F5 to launch a new VSCode window (Extension Development Host)
+3. Open a ReScript project to test the extension
+
+## Key Files
+
+### Configuration
+
+- `package.json`: Extension manifest, commands, settings, and scripts
+- `rescript.configuration.json`: Editor configuration for ReScript files
+- `client/src/extension.ts`: Extension entry point and client initialization
+- `server/src/server.ts`: Language server implementation
+- `server/src/cli.ts`: CLI entry point for the language server
+
+### OCaml Components
+
+- `analysis/`: Code analysis binary (hover, autocomplete, etc.)
+- `tools/`: ReScript tools binary (interface generation, etc.)
+
+### Language Features
+
+- **LSP Features**: hover, goto definition, find references, rename, autocomplete
+- **Code Analysis**: dead code detection, exception analysis (via reanalyze)
+- **Build Integration**: compile diagnostics, status indicators
+- **Commands**: interface creation, file switching, compiled JS opening
+
+## Testing
+
+The project has several test suites:
+
+- `analysis/tests/`: Tests for the analysis binary
+- `tools/tests/`: Tests for ReScript tools
+- `analysis/tests-incremental-typechecking/`: Incremental typechecking tests
+- `analysis/tests-generic-jsx-transform/`: JSX transformation tests
+
+## Project Structure Notes
+
+- The extension supports both `.res` (implementation) and `.resi` (interface) files
+- Uses VSCode Language Client protocol for communication between client and server
+- Native binaries are cross-platform (darwin, linux, win32) and included in the extension
```

Comment:
Maybe add a note about how this is actually working.
The `rescript-editor-analysis` is invoked by the LSP Server in a one-shot mode. Dumping JSON to stdout and the LSP picks that up.